File: 503. Next Greater Element II.cpp
class Solution {
public:
    // two pass stack: O(N)time & space, 56%, ok
    vector<int> nextGreaterElements(vector<int>& nums) {
        vector<int> ret(nums.size(), -1);  // initialize
        if(nums.empty()) return ret;
        stack<pair<int, int>> next;     // format: (val, index)
        int loop = 2;
        while(loop--){
            for(int i = 0; i < nums.size(); i++){
                while(next.size() && next.top().first < nums[i]){
                    int index = next.top().second;
                    if(ret[index] == -1)
                        ret[index] = nums[i];
                    next.pop();
                }
                next.push({nums[i], i});
            }
        }
        // set -1 for the largest value in vector
        int maxIdx = -1;
        while(next.size()){
            maxIdx = next.top().second;
            next.pop();
        }
        ret[maxIdx] = -1;
        return ret;
    }

    // more elegant: use stack to keep the indexes of the decreasing subsequence
    // O(N)time & space, 88.26%, ok
    vector<int> nextGreaterElements(vector<int>& nums) {
        vector<int> ret(nums.size(), -1);  // initialize
        if(nums.empty()) return ret;
        stack<int> next;
        for(int i = 0; i < nums.size() * 2; i++){
            int num = nums[i % nums.size()];
            while(next.size() && nums[next.top()] < num){
                ret[next.top()] = num;
                next.pop();
            }
            if(i < nums.size())   // key: key track of index
                next.push(i);
        }
        return ret;
    }
};

File: 503. Next Greater Element II.py
class Solution(object):
    
    # duplicate the list and brute force: O(n2) time, O(n) space, TLE
    def nextGreaterElements(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        nums1 = nums + nums
        ret = [-1 for i in xrange(len(nums))]
        for i in xrange(len(nums)):
            for j in xrange(i + 1, len(nums1)):
                if nums[i] < nums1[j]:
                    ret[i] = nums1[j]
                    break
        return ret
    
    # stack: O(n) time, O(n) space
    def nextGreaterElements(self, nums):
        ret, stack = [-1 for num in nums], []
        for i, num in enumerate(nums):
            while stack and stack[-1][1] < num:
                ret[stack.pop()[0]] = num
            stack.append((i, num))
        for num in nums:
            if len(stack) == 0:
                break
            while stack and stack[-1][1] < num:
                ret[stack.pop()[0]] = num
        return ret
                    
        